### ISSUE TITLE
SE: Preserve captures in Finally

### ISSUE THREAD

**Issue by pavel-mikula-sonarsource** [SonarSource/sonar-dotnet#7339]
The `Finally` block transition loses captures, because it's leaving that block.

We need to preserve them and release them after we leave the finally.

This should fix S3966 FN on the new engine.
```
    class Program
    {
        public void Disposed_Using_WithExpressions()
        {
            var d = new Disposable();
            using (d) // FN
            {
                d.Dispose();
            }
        }
    }

    public class Disposable : IDisposable
    {
        public void Dispose() { }
    }
```

As part of this, we also need to make sure that FlowReferenceCaptureOperation can track the state from the operation AND the symbol, if present.

**Comment by pavel-mikula-sonarsource** [comment]
If we'd make `FlowCaptureReference` and update the symbol releasing as it was in #7351, this could be fixed to support
```
using (d) // FN
```
and
```
using (s ??= new Disposable()) // FN
{
    s.Dispose();
}
```

The problem is that changes in releasing state break too many UTs and desired behaviors, because LVA is not smart enough around flow reference captures and try/catch/finally situations.

**Comment by pavel-mikula-sonarsource** [comment]
Isn't this issue about not loosing that information on finally?

**Comment by Tim-Pohlmann** [comment]
Yes, we have unnecessary branching, because we lose information that we should persist.

**Comment by pavel-mikula-sonarsource** [comment]
Yes, that's exactly what the original issue says that we should fix.

**Comment by pavel-mikula-sonarsource** [comment]
Internal ticket [NET-1774](https://sonarsource.atlassian.net/browse/NET-1774)

[NET-1774]: https://sonarsource.atlassian.net/browse/NET-1774?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

**Comment by pavel-mikula-sonarsource** [comment]
Moved
